fmt/format.h

```diff
@@ -2593,11 +2687,8 @@ class SystemError : public internal::RuntimeError {
        throw fmt::SystemError(errno, "cannot open file '{}'", filename);
    \endrst
   */
-  SystemError(int error_code, CStringRef message) {
-    init(error_code, message, ArgList());
-  }
```

Comment:
Please keep this ctor.

Reply:
It is generated by the FMT_VARIADIC_ macro.

Reply:
It's here for documentation purposes (note the apidoc above). If it is causing problems for bcc, I suggest conditionally compiling instead of removing.

Reply:
It's not a bcc issue, it's because all `FMT_VARIADIC_*` macros now use the `FMT_VARIADIC_` macro, this normalizing variadic template handing, and less macro magic to maintain. The `FMT_VARIADIC_` macro generates functions with 0..15 templated arguments to mimic variadic templates, where the 0 argument one has the exact same signature as this manually defined one.
I propose to make it consistent with functions `printf`, `format` etc and add an overload with an extra `ArgList` parameter:
```cpp
SystemError(int error_code, CStringRef message, ArgList args) {
    init(error_code, message, args);
  }
```
This would be the best of both worlds: singe `FMT_VARIADIC_` macro and documentation (and more consistency as a bonus :-) ).

Reply:
If this is not bcc issue, could you remove the macro changes from the current PR?

Reply:
The need for the macro changes *is* because of a bcc issue. If you feel very strongly about not changing the constructor signature in the documentation I can change the `FMT_VARIADIC_CTOR` macro not to generate the conflicting overload.

### Background

The original macro did c-style array initialization:
```cpp
ArgArray<2>::Type array = { MakeValue<BasicFormatter<char>>(v0), 
                            MakeValue<BasicFormatter<char>>(v1)};
```
Where `ArgArray<2>::Type` is an array of `Value`.
Somehow bcc tried to assign the results of `MakeValue<BasicFormatter<char>>(vN)` to the first member of `Value` (`int int_value` in the anonymous union) causing a "Cannot convert `MakeValue<BasicFormatter<char>>(vN) to int" error.

When I started fixing this issue I found the `FMT_VARIADIC_` function and started using that. This had already solved this issue by generating the individual assignment statements for the array:
```cpp
ArgArray<2>::Type arr;
arr[0] = MakeValue<BasicFormatter<char>>(v0), arr[1] = MakeValue<BasicFormatter<char>>(v1);
```

note: I left out namespaces for clarity

Reply:
I see. The macro logic is fairly complicated and version 4.x is in maintenance mode, so I'd like the changes in this PR to be minimal, just enough for bcc32 compatibility. Right now it's hard to see what actually changed because the macros have been moved around. Could you return them to their original places and make changes there?

Reply:
I started doing that, but found that the solution would depend on the (newer and more versatile) macros defined later on in the file. That is why I moved the large macros block up. To verify that I only moved them I used kdiff3 with manual alignment. I'll post those diffs so you can get a better idea of the changes I made.

Reply:
Here you can see what I changed  to get this working for bcc with the existing macros. Not only achieving bcc compatibility, but also removing the second set of macro's, resulting in less "magic" to maintain:
![Macro-diff](https://user-images.githubusercontent.com/944388/54927039-bae32f00-4f11-11e9-824a-1f04dd62256e.png)

And the diff for the `FMT_MAKE_REF` macro's:
![Macro-diff - old macros](https://user-images.githubusercontent.com/944388/54927106-d51d0d00-4f11-11e9-91f4-552890da9afd.png)

The alignment of the diff is shown below. I marked the manual alignment markers with red. 
![Macro-diff alignment](https://user-images.githubusercontent.com/944388/54927007-ae5ed680-4f11-11e9-95b5-019e1e431cd2.png)

I hope this makes it clear that I only made minor changes, apart from moving the macros.